File: JavaScript/080_removeDuplicates.js
/**
 * @param {number[]} nums
 * @return {number}
 * 
 * Runtime: 92 ms
 * Memory Usage: 40.5 MB
 * Big-O: O(n)
 */
var removeDuplicates = function(nums) {
  let isRepeated = false;
  for (let i=0; i<nums.length; i++) {
    // 如果數字已經重複兩次以上，則略過不計算
    if (isRepeated && nums[i] === nums[i-1]) {
      nums.splice(i, 1);
      i -= 1;
    }
    // 如果數字還未重複過但和前一個一樣，則isRepeated = true
    if (nums[i] === nums[i-1]) {
      isRepeated = true;
    } else {
      isRepeated = false;
    }
  }
  return nums.length;
};

/**
 * @param {number[]} nums
 * @return {number}
 * 
 * Runtime: 116 ms
 * Memory Usage: 40.6 MB
 * Big-O: unknown
 * Idea: 快慢指針＋計數器（血頭是最後的寫入），慢指針負責寫，快指針負責讀
 */
var removeDuplicates = function(nums) {
  // 因為題目表示相同的數字最多出現兩次，因此先過濾掉length<=2的陣列
  if (nums.length <= 2) {
    return nums.length;
  }
  // 建立快慢指針和計數器
  let slow = 0; // 寫頭
  let fast = 1; // 讀頭
  let count = 1;
  while (fast < nums.length) {
    // 兩指針的數字不同，則將獨到的數字寫在讀投下一個位置並歸零計數器
    if (nums[slow] !== nums[fast]) {
      nums[++slow] = nums[fast];
      count = 1;
    } else {
      // 兩指針的數字相同，則查看計數器是否已達上限
      if (count < 2) {
        // 計數器未達兩次則寫頭繼續往下紀錄，否則寫頭不動
        nums[++slow] = nums[fast];
        count ++;
      }
    }
    fast ++;
  }
  return slow + 1;
};

/**
 * @param {number[]} nums
 * @return {number}
 * 
 * Runtime: 132 ms	
 * Memory Usage: 40.5 MB
 * Big-O: unknown
 * Idea: 快慢指針（寫頭可直接寫入），慢指針負責寫，快指針負責讀
 */
var removeDuplicates = function(nums) {
  // 因為題目表示相同的數字最多出現兩次，因此先過濾掉length<=2的陣列
  if (nums.length <= 2) {
    return nums.length;
  }
  // 依據題目的需求建立慢指針 (countMax = 2)
  let slow = 2;
  for (let fast = 2; fast < nums.length; fast++) {
    if (nums[slow-2] !== nums[fast]) {
      nums[slow] = nums[fast];
      slow ++;
    }
  }
  return slow;
};

console.log(removeDuplicates([1,1,1,2,2,3]))
console.log(removeDuplicates([0,0,1,1,1,1,2,3,3]))

File: Python/080_removeDuplicates.py
from tkinter.messagebox import NO
from typing import List

class Solution1:
    def removeDuplicates(self, nums: List[int]) -> int:
        
        # Return answer if length less than 2
        if len(nums) <= 2: return len(nums)

        # While three elements are identical in a row
        i = 2
        while i < len(nums):
            if nums[i] == nums[i-1] == nums[i-2]:
                nums.pop(i)
            else:
                i += 1

        return i

class Solution2:
    '''
    解法貳利用快慢指針，並利用慢指針寫入，快指針做遍歷讀取
    '''
    def removeDuplicates(self, nums: List[int]) -> int:
        
        i = 0

        # 快指針透過遍歷掃過所有元素
        for num in nums:
            if i < 2 or num != nums[i-2]:
                # 若符合條件透過慢指針寫入
                nums[i] = num
                i += 1

        return i

class Solution3:
    '''
    解法三一樣使用快慢指針，差別是有明確訂出兩指針位置
    '''
    def removeDuplicates(self, nums: List[int]) -> int:

        n = len(nums)
        
        if n <= 2: return n

        count = 1
        slow, fast = 0, 1
        while fast < n:
            if nums[slow] != nums[fast]:
                slow +=1 
                nums[slow] = nums[fast]
                count = 1
            else:
                if count < 2:
                    slow +=1 
                    nums[slow] = nums[fast]
                    count += 1
            fast += 1

        return slow + 1


if __name__ == '__main__':
    sol = Solution3()
    ans1 = sol.removeDuplicates([1,1,1,2,2,3])
    print("Ans 1 is {}".format(ans1))
    ans2 = sol.removeDuplicates([0,0,1,1,1,1,2,3,3])
    print("Ans 2 is {}".format(ans2))
    ans3 = sol.removeDuplicates([1,1,1,1])
    print("Ans 3 is {}".format(ans3))